ws-security-dom/src/main/java/org/apache/wss4j/dom/message/WSSecEncryptedKey.java

```diff
@@ -561,6 +581,56 @@ private KeyAgreementParameters buildKeyAgreementParameter(PublicKey recipientPub
         return dhSpec;
     }
 
+    /**
+     * Method builds the KeyDerivationParameters for keyDerivationMethod and default values. The default values for
+     * the key derivation method are:
+     * <ul>
+     *   <li>ConcatKDF
+     *     <ul>
+     *       <li> DigestAlgorithm: "http://www.w3.org/2001/04/xmlenc#sha256"</li>
+     *       <li> AlgorithmID: "0000"</li>
+     *       <li> PartyUInfo: ""</li>
+     *       <li> PartyVInfo: ""</li>
+     *       <li> SuppPubInfo: null</li>
+     *       <li> SuppPrivInfo: null</li>
+     *     </ul>
+     *   <li>HKDF: SHA-256
+     *     <ul>
+     *       <li> PRF: http://www.w3.org/2001/04/xmldsig-more#hmac-sha256 </li>
+     *       <li> Salt: random 256 bit value</li>
+     *       <li> Info: null</li>
+     *     </ul>
+     *   </li>
+     * </ul>
+     *
+     * @param keyBitLength the length of the derived key in bits
+     * @return KeyDerivationParameters the {@link KeyDerivationParameters} for generating the
+     * key for encrypting transport key and generating XML elements.
+     * @throws WSSecurityException if the KeyDerivationParameters cannot be created
+     */
+    private KeyDerivationParameters buildKeyDerivationParameters(int keyBitLength) throws WSSecurityException {
+
+        switch (keyDerivationMethod) {
+            case WSS4JConstants.KEYDERIVATION_CONCATKDF:
+                return XMLCipherUtil.constructConcatKeyDerivationParameter(keyBitLength, WSConstants.SHA256,
+                        "0000", "", "", null, null);
+            case WSS4JConstants.KEYDERIVATION_HKDF:
+                // use semi random value for salt.
+                // rfc5869: Yet, even a salt value of less quality (shorter in
+                //   size or with limited entropy) may still make a significant
+                //   contribution to the security of the output keying material
+                byte[] semiRandom = new byte[keyBitLength / 8];
+                new Random().nextBytes(semiRandom);
```

Comment:
Shouldn't we use SecureRandom at least?

Reply:
In this particular case the SecureRandom does not significantly enhance cryptographic security but does impact performance. 
The Random is most optimal choice for most use cases with HKDF. 
However, if a user requires stronger cryptographic security for HKDF salt, they can provide their own value using keyDerivationParameters.”

Reply:
Can you expand a bit on how using Random is not a security issue?

Reply:
Sure.
As you can see from  rfc5869:  [3.1](https://datatracker.ietf.org/doc/html/rfc5869#section-3.1).  To Salt or not to Salt
 > Yet, even a salt value of less quality (shorter in
 >  size or with limited entropy) may still make a significant
 >  contribution to the security of the output keying material.
 
In this case (for the HKDF), the salt is public, and it doesn't matter if it's made from high or limited entropy, but it makes a difference as long as it's there.

Reply:
Could we make it configurable? e.g. default to SecureRandom but allow Random to be specified?

Reply:
I already made and configuration option for this (to use provided salt)
EncryptionTest.testEncryptionDecryptionWithKeyAgreementAndHKDF
and I have a suggestion then to make it to use by default  SecureRandom.
And if someone would like to improve  performance (for example with a systems with a high message load)  with KeyAgreement Method it can set its own parameter values using the option:
 builder.setKeyDerivationParameters(keyDerivationParameters);
 (or by setting the: ENC_KEY_DERIVATION_PARAMS)
 
 I do not think additional configuration option is needed then?

Reply:
Yes that's fine, if you can just switch Random to SecureRandom by default

Reply:
Done